### PR TITLE
fixes #14666 - support IPv6 in UI

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

File: app/views/nic/_base_form.html.erb [hunk 1]
```diff
@@ -28,25 +30,47 @@
...
               :help_inline => popover('',
-                  _("An IP address will be auto-suggested if you have a DHCP-enabled Smart Proxy on the subnet selected above.<br/><br/>The IP address can be left blank when:<br/><ul><li>provisioning tokens are enabled</li><li>the domain does not manage DNS</li><li>the subnet does not manage reverse DNS</li><li>and the subnet does not manage DHCP reservations</li></ul>"),
+                  _("An IP address will be auto-suggested if you have enabled IPAM on the IPv4 subnet selected above.<br/><br/>The IP address can be left blank when:<br/><ul><li>provisioning tokens are enabled</li><li>the domain does not manage DNS</li><li>the subnet does not manage reverse DNS</li><li>and the subnet does not manage DHCP reservations</li></ul>"),
                   :title => _("IP address auto-suggest"),
                   :rel => 'popover-modal',
                   :'data-placement' => "top"
                 ).html_safe,
                 :size => "col-md-8", :label_size => "col-md-3" %>
 
+<%= text_f f, :ip6,
```
Comment (dLobatog) [inline app/views/nic/_base_form.html.erb:53]: When I try to click on 'suggest new ip' , it returns 404 because I have not set a MAC yet. The error here should be a bit more informative than just 'not found'

Comment (timogoebel) [inline app/views/nic/_base_form.html.erb:53]: Note to self: When the mac is updated, a new ip is generated. We should maybe validate the mac and show errors early.

**Comment by timogoebel** [comment]
Rebased.

File: app/views/nic/_base_form.html.erb (hunk 1 above)
Comment (timogoebel) [inline app/views/nic/_base_form.html.erb:53]: @dLobatog : Implemented this. It still lacks a way what to do when the mac is generated by the compute resource.

File: app/views/subnets/_fields.html.erb [hunk 2]
```diff
@@ -1,20 +1,25 @@
 <%= text_f f, :name %>
+<%= selectable_f f, :type,
+  options_for_select(Subnet.types_with_form_data, @subnet.type),
+  {:label => _("Protocol")},
+  :disabled => !f.object.new_record? %>
 <%= text_f f, :network %>
-<%= text_f f, :mask, :help_inline => _("Netmask for this subnet") %>
+<%= text_f f, :cidr, :label => _("CIDR"), :help_inline => _("CIDR for this subnet") %>
```
Comment (dLobatog) [inline app/views/subnets/_fields.html.erb]: This is the "prefix length", not the CIDR, the CIDR is `fc00:0:0:55::/64`, but this field refers to the 64 part, thus prefix length.

File: app/views/subnets/_fields.html.erb [hunk 3]
```diff
@@ -1,20 +1,25 @@
 <%= text_f f, :name %>
+<%= selectable_f f, :type,
+  options_for_select(Subnet.types_with_form_data, @subnet.type),
+  {:label => _("Protocol")},
```
Comment (dLobatog) [inline app/views/subnets/_fields.html.erb]: I think you mean:

``` ruby
{ }, 
:label => _("Protocol"), :disabled...
```

Otherwise the label is ignored and it shows Type

File: app/views/nic/_base_form.html.erb [hunk 4]
```diff
@@ -28,25 +30,47 @@
...
+              :label => _("IPv6 address"),
+              :class => :interface_ip6,
+              :autocomplete => 'off',
+              :help_block => link_to(_("Suggest new"), '#', :class => 'suggest_new_ip6'),
```
Comment (dLobatog) [inline app/views/nic/_base_form.html.erb]: This `Suggest new` should only be shown when that's actually possible: which is when `subnet.ipam` is `Internal DB`. Otherwise one can click w/o any feedback and it's confusing

File: app/services/ipam/dhcp.rb [hunk 5]
```diff
@@ -8,6 +8,10 @@ def suggest_ip
       ip = dhcp_proxy.unused_ip(subnet, mac)["ip"]
       logger.debug("Found #{ip}")
       ip
+    rescue => e
+      logger.warn "Failed to fetch a free IP from our proxy: #{e}"
```
Comment (dLobatog) [inline app/services/ipam/dhcp.rb]: Can you also add the proxy name and URL in here? It's a bit of a pain to have to click on subnet and check what proxy is associated with the subnet when you're creating a host

File: app/views/subnets/_fields.html.erb (hunk 2 above)
Comment (timogoebel) [inline app/views/subnets/_fields.html.erb]: @dLobatog: To be honest I don't want to argue about this. But If I'm not mistaken "prefix length" is an IPv6 term, for IPv4 it's called "suffix". CIDR just says, that you don't specify the subnet mask in dotted decimal notation, but you count the number of 1-bits in the subnet mask. That's why I chose CIDR, but I don't mind changing this.
I guess it would be better to show the subnet mask in dotted decimal notation for IPv4 and the prefix length for IPv6. That would require us to filter the opposite param in the controller. Do you think, that's worth the effort? Is there a better approach?

File: app/helpers/hosts_and_hostgroups_helper.rb [hunk 6]
```diff
@@ -48,8 +50,8 @@ def accessible_puppet_proxies(obj)
     list
   end
 
-  def domain_subnets
-    accessible_related_resource(@domain, :subnets)
+  def domain_subnets(type)
+    accessible_related_resource(@domain, :subnets, :query => {:type => type})
```
Comment (tbrisker) [inline app/helpers/hosts_and_hostgroups_helper.rb]: Why not `accessible_related_resource(@domain, :subnets).where(:type=> type)` and then there is no need for changing the `accessible_related_resource` function?

Comment (timogoebel) [inline app/helpers/hosts_and_hostgroups_helper.rb]: @tbrisker: I couldn't do it because `return [] if obj.blank?` would lead to `[].where(:type => type)` and that obviously does not work. Is there a better approach?

**Comment by timogoebel** [comment]
@dLobatog : Addressed your comments. Added hiding of the "suggest new links" when the subnet does not support suggesting a new ip.

File: app/helpers/hosts_and_hostgroups_helper.rb (hunk 6 above)
Comment (tbrisker) [inline app/helpers/hosts_and_hostgroups_helper.rb]: aha, I see. that makes sense, could you please only rename the option from `:query` to `:where` so it is clear this is passed as an additional `where` to the filter? Also, please extract the options from the hash only after the `blank?` short circuit, no point in extracting them if they won't be used.

Comment (timogoebel) [inline app/helpers/hosts_and_hostgroups_helper.rb]: @tbrisker : Sure, done. 👍

**Comment by dLobatog** [comment]
@timogoebel I'm a bit confused about the CIDR vs prefix length discussion in https://github.com/theforeman/foreman/pull/3587#discussion_r69083574 . What did you want that field to contain? You can also use a subnet mask in dotted hexadecimal notation for IPv6 but I don't think it's common vs using a suffix. 

`That would require us to filter the opposite param in the controller. Do you think, that's worth the effort? Is there a better approach?`

This bit confused me too, what's 'the opposite param in the controller'?

**Comment by timogoebel** [comment]
@dLobatog 

> I'm a bit confused about the CIDR vs prefix length discussion in #3587 (comment) .

Sorry for not making this clear. I meant to say the following:
- Currently we use the subnet mask for IPv4 (255.255.255.0), which is fine.
- For IPv6 the subnet mask is not used as you said (ffff:ffff:ffff:ffff:0000:0000:0000:0000)
- The current status of this patch changes both IPv4 and IPv6 to use the "cidr"/prefix ("24" as in 192.168.1.1/24; "64" as in 2001:db8::1/64)
- We could change the UI to use "255.255.255.255" for IPv4 and "64" for IPv6.

My first thought on how to implement the latter is to just toggle visibility of the prefix and the mask field via jQuery. The browser than still sends both information in the request. That's why I meant we need to filter it in the controller. There might be a better approach, haven't come up with something though.

File: app/services/ipam/base.rb [hunk 7]
```diff
@@ -1,23 +1,40 @@
...
 
     def normalize_mac!
       return unless self.mac.present?
       self.mac = Net::Validations.normalize_mac(self.mac)
     rescue Net::Validations::Error
-      raise Foreman::Exception.new(N_("'%s' is not a valid MAC address.") % self.mac)
+      errors.add(:mac, _('is not a valid MAC address'))
+    end
+
+    def read_attribute_for_validation(attr)
```
Comment (dLobatog) [inline app/services/ipam/base.rb:39]: I had to look up what this does, mind to add a comment explaining why this is ncessary?

File: app/services/ipam/base.rb [hunk 8]
```diff
@@ -1,23 +1,40 @@
...
+      @errors = ActiveModel::Errors.new(self)
 
       normalize_mac!
     end
 
+    def suggest_new?
+      true
+    end
+
+    def serializable_hash(options = nil)
```
Comment (dLobatog) [inline app/services/ipam/base.rb:25]: When is this serialized?

File: app/services/ipam/base.rb [hunk 9]
```diff
@@ -1,23 +1,40 @@
 module IPAM
   class Base
+    extend ActiveModel::Naming
```
Comment (dLobatog) [inline app/services/ipam/base.rb:4]: Sorry, what are these two for?

File: app/assets/javascripts/host_edit.js [hunk 10]
```diff
@@ -550,40 +544,70 @@ $(document).on('change', '#host_provision_method_image', image_provision_method_
...
 
 $(document).on('click', '.suggest_new_ip', function (e) {
-  clearIpField(this);
-  interface_subnet_selected($(this).closest('fieldset').find('select.interface_subnet'));
+  clearIpField(this, '.interface_ip');
+  interface_subnet_selected($(this).closest('fieldset').find('select.interface_subnet'), 'ip');
+  e.preventDefault();
+});
+
+$(document).on('click', '.suggest_new_ip6', function (e) {
```
Comment (dLobatog) [inline app/assets/javascripts/host_edit.js:570]: Duplicate code, it's same as ipv4 except for 1 parameter

File: app/assets/javascripts/host_edit.js [hunk 11]
```diff
@@ -550,40 +544,70 @@ $(document).on('change', '#host_provision_method_image', image_provision_method_
...
   e.preventDefault();
 });
 
 $(document).on('change', '.interface_subnet', function () {
-  clearIpField(this);
-  interface_subnet_selected(this);
+  interface_subnet_selected(this, 'ip');
+});
+
+$(document).on('change', '.interface_subnet6', function () {
```
Comment (dLobatog) [inline app/assets/javascripts/host_edit.js:578]: Duplicate code

File: app/assets/javascripts/subnets.js [hunk 12]
```diff
@@ -1,10 +1,43 @@
...
 
+function showSubnetProxies(element) {
+  if (element.find(':selected').data('supports_dhcp')) {
+    $('#subnet_dhcp_id').closest('.form-group').show();
+  } else {
+    $('#subnet_dhcp_id').closest('.form-group').hide();
+    $("#subnet_dhcp_id option[value='']").attr("selected","selected");
+  }
+
+  if (element.find(':selected').data('supports_tftp')) {
```
Comment (dLobatog) [inline app/assets/javascripts/subnets.js]: duplicate code (dhcp/tftp)

File: app/models/subnet/ipv6.rb [hunk 13]
```diff
@@ -26,6 +26,10 @@ def self.supported_ipam_modes
     [:eui64, :db, :none]
   end
 
+  def self.supports_tftp?
+    false
```
Comment (dLobatog) [inline app/models/subnet/ipv6.rb:30]: Why is tftp not supported? xinetd / tftpd which come with the foreman installer support IPv6

File: app/services/ipam/eui64.rb [hunk 14]
```diff
@@ -1,16 +1,27 @@
...
+    def suggest_new?
+      false
+    end
+
     private
 
+    def validate
+      errors.add(:mac, _("can't be blank")) unless mac.present?
+      errors.add(:subnet, _("network can't be blank")) unless subnet.network.present?
+      errors.add(:subnet, _("CIDR can't be blank")) unless subnet.cidr.present?
```
Comment (dLobatog) [inline app/services/ipam/eui64.rb]: "did you mean prefix length?" - clippy

File: .rubocop_todo.yml [hunk 15]
```diff
@@ -89,7 +89,7 @@ Metrics/MethodLength:
 # Offense count: 21
 # Configuration parameters: CountComments.
 Metrics/ModuleLength:
-  Max: 408
+  Max: 431
```
Comment (dLobatog) [inline .rubocop_todo.yml]: Can you find some stuff to split out in the helper so that it doesn't go beyond the limit? https://github.com/theforeman/foreman/pull/3621/files I had to do that here , if that works for you feel free to take it

**Comment by dLobatog** [comment]
@timogoebel 

> ```
> The current status of this patch changes both IPv4 and IPv6 to use the "cidr"/prefix ("24" as in 192.168.1.1/24; "64" as in 2001:db8::1/64) 
> ```

That's only because the field that used to contain 'mask' now it's set to 'cidr'. What you said before about toggling visibility of the prefix and mask makes sense IMO. Especially since the mask field is used **everywhere** in Foreman. 

The way I envision it is:

ipv4 - displays mask, cidr field is ignored (or set from the mask?, just trying to make it easy for now)
ipv6 - displays prefix, mask field is ignored 

Makes sense?

File: app/models/subnet/ipv6.rb (hunk 13 above)
Comment (timogoebel) [inline app/models/subnet/ipv6.rb:30]: They do, but Foreman currently does not support this. IMHO tftp orchestration only makes sense in combination with dhcp orchestration. We need efi support in Foreman first before dhcp orchestration can work. So this is still a lot more work to do.

File: app/services/ipam/base.rb (hunk 9 above)
Comment (timogoebel) [inline app/services/ipam/base.rb:4]: They are required for Active Model Errors, see http://api.rubyonrails.org/classes/ActiveModel/Errors.html

File: app/services/ipam/base.rb (hunk 8 above)
Comment (timogoebel) [inline app/services/ipam/base.rb:25]: https://github.com/timogoebel/foreman/blob/cdb33fed2b996e6e5fa4d6b610e5ff3a7a449260/app/controllers/concerns/foreman/controller/host_details.rb#L27-L29

`#domain_selected`  in `app/controllers/concerns/foreman/controller/host_details.rb` calls `as_json`, this needs the `serializable_hash` method.

File: app/assets/javascripts/host_edit.js (hunk 11 above)
Comment (timogoebel) [inline app/assets/javascripts/host_edit.js:578]: Any good idea on how to refactor this out?

**Comment by timogoebel** [comment]
> The way I envision it is:
> 
> ipv4 - displays mask, cidr field is ignored (or set from the mask?, just trying to make it easy for now)
> ipv6 - displays prefix, mask field is ignored

@dLobatog : Addressed most of your comments. Your vision became reality. :-P Had to add some client side validation for this to work. Mind to test the subnet form? Happy to hear your thoughts.

File: app/models/subnet/ipv6.rb (hunk 13 above)
Comment (dLobatog) [inline app/models/subnet/ipv6.rb:30]: Sorry, I still don't understand, TFTP orchestration is independent of DHCP, in fact some people do static network provisioning with Foreman (http://serverfault.com/questions/609770/using-foreman-to-provision-a-vm-with-static-ip for instance). The EFI part caught me completely off guard, why is that necessary?

Comment (timogoebel) [inline app/models/subnet/ipv6.rb:30]: @dLobatog : Just a short answer for now, don't have a lot of time today:

For tftp orchestration to work, we need to make changes to `app/models/concerns/orchestration/tftp.rb` before it can work for ipv6. There should be "subnet6" somewhere, haven't given it much thought right now. My main problem is that it's actually IPv4 and IPv6 independent. So when a host has an IPv4 subnet and a IPv6 subnet and both have a tftp proxy set, which one do you choose. That's why I wanted to hide it in the UI for now. We can easily add it later again when it's properly tested.

To provision a server with ipv6 only you would need the bios to make a dhcpv6 request. Most bioses don't support this. Most EFI firmwares do though. That's why I wrote we need EFI support. For VMWare you have to tell the compute resource to do an EFI boot and explicity set it to boot via IPv6. fog-vsphere doesn't support this right now. So we'd (read: I'd) have to add this first.

Does that make sense to you? If you want, we can do a hangout and discuss things there.

File: bundler.d/assets.rb [hunk 16]
```diff
@@ -16,3 +16,4 @@
 gem 'spice-html5-rails', '~> 0.1.5'
 gem 'quiet_assets', '~> 1.0'
 gem 'jquery_pwstrength_bootstrap_4', :require => 'jquery_pwstrength_bootstrap'
+gem 'ipaddrjs-rails', '~> 1.1.1'
```
Comment (dLobatog) [inline bundler.d/assets.rb:19]: This needs to be packaged, @timogoebel let me know if you want me to do it or you prefer to do it. The README in https://github.com/theforeman/foreman-packaging/tree/rpm/develop explains how to do this, it's very simple (run gem2rpm , update comps/ and rel-eng/tito.props and submit the PR :smile: )

**Comment by dLobatog** [comment]
@timogoebel Thanks, the latest updates look great. I can't find anything I don't agree with, or that doesn't work on the UI, so I'm merging it. Please respond to https://github.com/theforeman/foreman/pull/3587#discussion_r70662494 :smile: 

Thanks again @timogoebel , it's been a lot of work!

**Comment by dLobatog** [comment]
Merged as 79c0664a008afe0fea54a0b5256426f3867a02bc, thanks @timogoebel!

File: bundler.d/assets.rb (hunk 16 above)
Comment (timogoebel) [inline bundler.d/assets.rb:19]: No worries, I'll do it tomorrow. :-)
